### ISSUE TITLE
upgrade to 7.0.0 leads to "stack level too deep" errors

### ISSUE THREAD

**Issue by gsar** [newrelic/newrelic-ruby-agent#668]
## Description

It appears newrelic_rpm v7.0.0 conflicts with the aliased method chaining used in [this spot](https://github.com/airbrake/airbrake/blob/dfbd162a6f9f6bdeee7eeb5e7d1449d1770091fb/lib/airbrake/rails/http.rb#L10) in airbrake v11.0.1

Perhaps this case needs to be added to the exception list included in newrelic?

## Expected Behavior

There should be no errors.

## [Troubleshooting](https://discuss.newrelic.com/t/troubleshooting-frameworks/108787) or [NR Diag](https://docs.newrelic.com/docs/using-new-relic/cross-product-functions/troubleshooting/new-relic-diagnostics) results

```
stack level too deep
/app/vendor/bundle/ruby/3.0.0/gems/addressable-2.7.0/lib/addressable/idna/pure.rb:305:in `lookup_unicode_composition'
/app/vendor/bundle/ruby/3.0.0/gems/addressable-2.7.0/lib/addressable/idna/pure.rb:215:in `unicode_compose_pair'
/app/vendor/bundle/ruby/3.0.0/gems/addressable-2.7.0/lib/addressable/idna/pure.rb:153:in `block in unicode_compose'
/app/vendor/bundle/ruby/3.0.0/gems/activesupport-6.1.3.2/lib/active_support/core_ext/range/each.rb:9:in `each'
/app/vendor/bundle/ruby/3.0.0/gems/activesupport-6.1.3.2/lib/active_support/core_ext/range/each.rb:9:in `each'
/app/vendor/bundle/ruby/3.0.0/gems/addressable-2.7.0/lib/addressable/idna/pure.rb:149:in `unicode_compose'
/app/vendor/bundle/ruby/3.0.0/gems/addressable-2.7.0/lib/addressable/idna/pure.rb:121:in `unicode_normalize_kc'
/app/vendor/bundle/ruby/3.0.0/gems/addressable-2.7.0/lib/addressable/uri.rb:566:in `normalize_component'
/app/vendor/bundle/ruby/3.0.0/gems/addressable-2.7.0/lib/addressable/uri.rb:879:in `normalized_scheme'
/app/vendor/bundle/ruby/3.0.0/gems/addressable-2.7.0/lib/addressable/uri.rb:2160:in `normalize'
/app/vendor/bundle/ruby/3.0.0/gems/addressable-2.7.0/lib/addressable/uri.rb:2184:in `normalize!'
/app/vendor/bundle/ruby/3.0.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/http_clients/uri_util.rb:35:in `parse_and_normalize_url'
/app/vendor/bundle/ruby/3.0.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/http_clients/http_rb_wrappers.rb:32:in `uri'
/app/vendor/bundle/ruby/3.0.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/httprb/instrumentation.rb:13:in `with_tracing'
/app/vendor/bundle/ruby/3.0.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/httprb/prepend.rb:11:in `perform'
/app/vendor/bundle/ruby/3.0.0/gems/airbrake-11.0.1/lib/airbrake/rails/http.rb:10:in `block in perform'
/app/vendor/bundle/ruby/3.0.0/gems/airbrake-11.0.1/lib/airbrake/rack.rb:25:in `capture_timing'
/app/vendor/bundle/ruby/3.0.0/gems/airbrake-11.0.1/lib/airbrake/rails/http.rb:9:in `perform'
/app/vendor/bundle/ruby/3.0.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/httprb/prepend.rb:11:in `block in perform'
/app/vendor/bundle/ruby/3.0.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/httprb/instrumentation.rb:19:in `block in with_tracing'
/app/vendor/bundle/ruby/3.0.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/tracer.rb:371:in `capture_segment_error'
/app/vendor/bundle/ruby/3.0.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/httprb/instrumentation.rb:19:in `with_tracing'
/app/vendor/bundle/ruby/3.0.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/httprb/prepend.rb:11:in `perform'
/app/vendor/bundle/ruby/3.0.0/gems/airbrake-11.0.1/lib/airbrake/rails/http.rb:10:in `block in perform'
/app/vendor/bundle/ruby/3.0.0/gems/airbrake-11.0.1/lib/airbrake/rack.rb:25:in `capture_timing'
/app/vendor/bundle/ruby/3.0.0/gems/airbrake-11.0.1/lib/airbrake/rails/http.rb:9:in `perform'
/app/vendor/bundle/ruby/3.0.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/httprb/prepend.rb:11:in `block in perform'
/app/vendor/bundle/ruby/3.0.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/httprb/instrumentation.rb:19:in `block in with_tracing'
/app/vendor/bundle/ruby/3.0.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/tracer.rb:371:in `capture_segment_error'
/app/vendor/bundle/ruby/3.0.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/httprb/instrumentation.rb:19:in `with_tracing'
/app/vendor/bundle/ruby/3.0.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/httprb/prepend.rb:11:in `perform'
/app/vendor/bundle/ruby/3.0.0/gems/airbrake-11.0.1/lib/airbrake/rails/http.rb:10:in `block in perform'
/app/vendor/bundle/ruby/3.0.0/gems/airbrake-11.0.1/lib/airbrake/rack.rb:25:in `capture_timing'
/app/vendor/bundle/ruby/3.0.0/gems/airbrake-11.0.1/lib/airbrake/rails/http.rb:9:in `perform'
/app/vendor/bundle/ruby/3.0.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/httprb/prepend.rb:11:in `block in perform'
/app/vendor/bundle/ruby/3.0.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/httprb/instrumentation.rb:19:in `block in with_tracing'
/app/vendor/bundle/ruby/3.0.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/tracer.rb:371:in `capture_segment_error'
/app/vendor/bundle/ruby/3.0.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/httprb/instrumentation.rb:19:in `with_tracing'
/app/vendor/bundle/ruby/3.0.0/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/httprb/prepend.rb:11:in `perform'
/app/vendor/bundle/ruby/3.0.0/gems/airbrake-11.0.1/lib/airbrake/rails/http.rb:10:in `block in perform'
/app/vendor/bundle/ruby/3.0.0/gems/airbrake-11.0.1/lib/airbrake/rack.rb:25:in `capture_timing'
/app/vendor/bundle/ruby/3.0.0/gems/airbrake-11.0.1/lib/airbrake/rails/http.rb:9:in `perform'
...etc...
```

## Steps to Reproduce
Please be as specific as possible.
TIP:  Link a sample application that demonstrates the issue. 

## Your Environment

See versions of packages involved in the stack trace above. Reproducible on ruby 3.0.1 and rails 6.1.3.2.

## Additional context
Add any other context about the problem here. For example, relevant community posts or support tickets.

**Comment by gsar** [comment]
https://github.com/airbrake/airbrake/issues/1161

**Comment by angelatan2** [comment]
Noted that airbrake has addressed the issue.  We are keeping this open until Airbrake has a release version.

**Comment by angelatan2** [comment]
@tannalynn, @amhuntsman Please note that Airbrake 11.0.3 fixes the issues that were generating stack level too deep (ie. conflicting monkey patching strategies between alias and prepend).  Please update our configuration to use prepend with airbrake gem for version 11.0.3 and above.

**Comment by gsaks123** [comment]
I ran into the stack level too deep issue when combining Airbrake 11.0.3 and New Relic 6.15.0. I raised an issue in the Airbrake repo, but it was closed:
https://github.com/airbrake/airbrake/issues/1171

At this point, what is the recommended version of New Relic to be using? When looking at the [release tags](https://github.com/newrelic/newrelic-ruby-agent/releases) here I see all the recent releases are still tagged with "prerelease".   And what is the New Relic prepend config that is being discussed, and which version should it be used with?

I could move my app to New Relic 7.x and Airbrake 11.0.3, and report back on how it goes. But each time I deploy and get the stack level too deep errors, I have tons of failed Resque jobs within minutes. I don't want to have to rollback and clean this up more times than needed. So, basically, I could give it a shot if it would help the teams here diagnose potential issues. On the other hand if there are still known issues being worked out, I'll just wait on the sidelines a bit longer.

**Comment by tannalynn** [comment]
@gsaks123
Thanks for letting us know about the issues you're experiencing.
So the `stack level too deep` errors are caused by gems that use both Module#prepend and method chaining. These are incompatible and cannot be used on the same methods, or a `stack level too deep` error will occur. In the ruby agent, prior to 7.0, we used method chaining on most of our instrumentation. The 7.0 release introduces using Module#prepend by default, but allows configuration options to choose to use method chaining in cases where you are using another gem that conflicts with the use of Module#prepend in our instrumentation.

Since airbrake has updated to use prepend in a few spots in their latest release, ruby agent 7.0+ should be used in combination with those newer airbrake versions to avoid the stack level too deep errors. 
The latest released version of the agent is 7.1.0. This is the [tag for the release](https://github.com/newrelic/newrelic-ruby-agent/tree/7.1.0). All releases are available on rubygems as well, of course. 

For upgrading to 7.0, I'd recommend reviewing the [migration guide](https://docs.newrelic.com/docs/agents/ruby-agent/getting-started/migration-7x-guide/) and our [new config options](https://docs.newrelic.com/docs/agents/ruby-agent/configuration/ruby-agent-configuration/#instrumentation) for instrumentation. If you run into a prepend conflict (which shows up as a `stack level too deep` error), update the config file to use chain for the instrumentation on the gem that the conflict occurs on. Also, by default the agent checks for known conflicts with prepend to choose chain instead when these conflicts are found. 
You mentioned failing resque jobs, We currently have a check on our resque instrumentation to check for airbrake since they were previously using chain. Since this latest release updates that, I've created a PR that checks for the airbrake version now, so that it will choose prepend when using airbrake 11.0.3+. This conflict checking can be bypassed by setting the desired value for instrumentation in the config file. So if the conflict is happening on the resque instrumentation when using auto (the default value), adding this to your newrelic.yml
```
instrumentation:
  resque: prepend
```
should bypass the prepend conflict check and use exactly the type of instrumentation you give. This prepend check for 'auto' will updated in our next release to include checking this new version.
 It's possible conflicts can be occurring in some other instrumentation used as well, depending on your environment and what gems you are using, so you may need to add additional config options to control the instrumentation used for other gems.

Hopefully that helps explain some of the changes related to the stack level too deep errors related to method chaining and Module#prepend.

**Comment by gsaks123** [comment]
Thanks @tannalynn for your detailed reply. I'll wait for a version that includes your PR and then give it a go!
